src/serialize.rs

```diff
@@ -0,0 +1,222 @@
+            NativeType::Boolean => self.serialize_native::<bool>(typ, writer),
+            NativeType::BigInt => self.serialize_native::<i64>(typ, writer),
+            NativeType::Double => self.serialize_native::<f64>(typ, writer),
+            NativeType::Int => self.serialize_native::<i32>(typ, writer),
+            NativeType::Text => self.serialize_native::<String>(typ, writer),
+            _ => unimplemented!("other native types not supported yet"),
+        }
+    }
+
+    fn serialize_native<'a, T>(
+        &self,
+        typ: &ColumnType,
+        writer: CellWriter<'a>,
+    ) -> Result<WrittenCellProof<'a>, SerializationError>
+    where
+        T: for<'b> pyo3::FromPyObject<'b> + SerializeValue,
+    {
+        let value = Python::with_gil(|py| {
+            self.bind(py)
+                .extract::<T>()
+                .map_err(SerializationError::new)
+        })?;
+
+        value.serialize(typ, writer)
+    }
+
+    fn serialize_types<'a>(
+        &self,
+        typ: &ColumnType,
+        writer: CellWriter<'a>,
+    ) -> Result<WrittenCellProof<'a>, SerializationError> {
+        match typ {
+            ColumnType::Native(native_type) => self.serialize_natives(typ, native_type, writer),
+            ColumnType::Collection {
+                frozen: _,
+                typ: collection_typ,
+            } => match collection_typ {
+                CollectionType::List(element_type) => {
+                    let list = PyListWrapper::new(self, typ).map_err(SerializationError::new)?;
+                    list.serialize(element_type, writer)
+                }
+                _ => unimplemented!("other collection types not supported yet"),
+            },
+            // Supports UDTs passed as Python dicts.
+            // For Python dataclass instances, convert to dict first (e.g., using dataclasses.asdict()).
+            ColumnType::UserDefinedType { definition, .. } => {
+                let dict = PyUdtWrapper::new(self, definition).map_err(SerializationError::new)?;
+                dict.serialize(typ, writer)
+            }
+            _ => unimplemented!("other types not supported yet"),
+        }
+    }
+}
+
+impl SerializeRow for PyAnyWrapper {
+    fn serialize(
+        &self,
+        ctx: &RowSerializationContext<'_>,
+        writer: &mut RowWriter,
+    ) -> Result<(), SerializationError> {
+        Python::with_gil(|py| {
+            let val = self.bind(py);
+
+            match val.downcast::<PyTuple>() {
+                Ok(py_tuple) => {
+                    for (col, val) in ctx.columns().iter().zip(py_tuple.iter()) {
+                        let wrapper = PyAnyWrapper(val.into());
+                        let sub_writer = writer.make_cell_writer();
+                        SerializeValue::serialize(&wrapper, col.typ(), sub_writer)?;
+                    }
+                    Ok(())
+                }
+                Err(_) => Err(invalid_input("expected Python tuple")),
```

Comment:
🤔 We're returning `SerializationError`, whereas this looks more like a Python `TypeError` to me (because the tuple is expected as the container for bound values).